Sink Island Check Enhancements

savannahostrowski opened the pull request:
The enhancements made to the Sink Island check filter out Ways from flagging candidates which contain an amenity tag with the value of either Parking, Parking Space, Motorcycle Parking, or Parking Entrance on the end node.

savannahostrowski commented:
@taylorulien If you have anything else to document for other enhancements, please write them here so I can keep track! 😄

savannahostrowski commented:
@danielduhh Ready for review.

danielduhh commented:
build error
```bash
[ant:checkstyle] [ERROR] /Users/danielbaah/Documents/github/atlas-checks/src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:205:52: Parameter object should be final. [FinalParameters]
[ant:checkstyle] [ERROR] /Users/danielbaah/Documents/github/atlas-checks/src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:205:64: Variable 'object' should be declared final. [FinalLocalVariable]
[ant:checkstyle] [ERROR] /Users/danielbaah/Documents/github/atlas-checks/src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:209:20: Variable 'tagValue' should be declared final. [FinalLocalVariable]
```

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:
```diff
@@ -180,4 +192,23 @@ private boolean validEdge(final AtlasObject object)
+     */
+    private boolean endNodeHasAmenityTypeToExclude(AtlasObject object)
+    {
+        if (((Edge) object).end().getTag(AmenityTag.KEY).isPresent())
```
`((Edge) object).end()` is called twice, could put into local variable

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:
```diff
@@ -180,4 +192,23 @@ private boolean validEdge(final AtlasObject object)
+            String tagValue = ((Edge) object).end().tag(AmenityTag.KEY);
+            return amenityValuesToExclude.contains(tagValue.toUpperCase());
+        }
+        return false;
```
I'm wondering if this can be consolidated into a single return statement

danielduhh replied:
Something like...

```java
        return tagValue.isPresent()
                && amenityValuesToExclude.contains(tagValue.get().toUpperCase());
```

savannahostrowski commented:
Added unit test and also added some functionality to make sure that we are returning `Optional.empty() `if no edges are found to flag after we filter for amenities.

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:
```diff
@@ -146,9 +154,19 @@ public boolean validCheckForObject(final AtlasObject object)
-                : Optional.of(createFlag(explored, this.getLocalizedInstruction(0)));
+        // We want to filter out any Edges whose end Node has an amenity tag equal to one of the
+        // amenityValuesToExclude
+        final Set<AtlasObject> filteredExploredEdges = explored.stream()
```
Should we filter out parking end nodes before they get marked as flagged?

savannahostrowski replied:
I think it's okay to leave it as it. I'm hesitant to mutate the list of explored (considering we still always want to mark all the edges we have seen, just not flag them). Ultimately, run time is still going to be O(n).

danielduhh commented on src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java:
```diff
@@ -33,4 +33,13 @@ public void testSinkDetection()
     }
+
+    @Test
+    public void loneEdgeWithAmenity()
```
rename to `testSingleEdgeWithAmenity` (just to be consistent with above functions)

danielduhh commented on src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTestRule.java:
```diff
@@ -59,6 +59,13 @@
+
+            edges = { @Edge(id = "360978519000003", coordinates = { @Loc(value = TEST_6),
+                    @Loc(value = TEST_7) }, tags = { "highway=primary", "oneway=yes" }) })
+    private Atlas singleEdgeAtlasWithAmenity;
```
rename to `singleEdgeWithAmenityAtlas` (for consistency)

danielduhh commented on src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTestRule.java:
```diff
@@ -68,4 +75,9 @@ public Atlas getTestAtlas()
         return this.testAtlas;
     }
+
+    public Atlas getSingleEdgeAtlasWithAmenity()
```
see above

danielduhh commented:
Changes approved 👊🏿
